### PR TITLE
Add new conf kyuubi.frontend.protocols

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

Adapt to multiple frontends protocol.

This PR also marks REST API as experimental and exposes it to users.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by pan3793** [comment]
cc @yanghua

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -254,6 +254,26 @@ object KyuubiConf {
...
+    val THRIFT_BINARY, REST = Value
+  }
+
+  val FRONTEND_PROTOCOLS: ConfigEntry[Seq[String]] =
+    buildConf("frontend.protocols")
+      .doc("A comma separated list for all frontend protocols " +
+        "<ul>" +
+        " <li>THRIFT_BINARY - HiveServer2 compatible thrift binary protocol.</li>" +
+        " <li>REST - Kyuubi defined REST API(experimental).</li> " +
+        "</ul>")
```
Comment (yanghua) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:268]: Shall we point out `THRIFT_BINARY ` is the default protocol? if users do nothing.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiServer.scala [hunk 2]
```diff
@@ -84,8 +86,15 @@ class KyuubiServer(name: String) extends Serverable(name) {
...
 
-  override val frontendServices: Seq[AbstractFrontendService] = Seq(
-    new KyuubiThriftBinaryFrontendService(this))
+  override lazy val frontendServices: Seq[AbstractFrontendService] =
+    conf.get(FRONTEND_PROTOCOLS).map(FrontendProtocols.withName).map {
+      case THRIFT_BINARY => new KyuubiThriftBinaryFrontendService(this)
+      case REST =>
+        warn("REST frontend protocol is experimental, API may changes in the future.")
+        new RestFrontendService(this)
+      case us =>
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiServer.scala:95]: what does the `us` mean? `unsupported`? `other` looks better?

Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiServer.scala:95]: yea, `us` means `unsupported`, I'm fine with `other`

File: docs/deployment/settings.md [hunk 3]
```diff
@@ -205,6 +205,7 @@ kyuubi\.frontend\.login<br>\.timeout|<div style='width: 65pt;word-wrap: break-wo
...
 kyuubi\.frontend\.min<br>\.worker\.threads|<div style='width: 65pt;word-wrap: break-word;white-space: normal'>9</div>|<div style='width: 170pt;word-wrap: break-word;white-space: normal'>(deprecated) Minimum number of threads in the of frontend worker thread pool for the thrift frontend service</div>|<div style='width: 30pt'>int</div>|<div style='width: 20pt'>1.0.0</div>
+kyuubi\.frontend<br>\.protocols|<div style='width: 65pt;word-wrap: break-word;white-space: normal'>THRIFT_BINARY</div>|<div style='width: 170pt;word-wrap: break-word;white-space: normal'>A comma separated list for all frontend protocols <ul> <li>THRIFT_BINARY - HiveServer2 compatible thrift binary protocol.</li> <li>REST - Kyuubi defined REST API(experimental).</li> </ul></div>|<div style='width: 30pt'>seq</div>|<div style='width: 20pt'>1.4.0</div>
```
Comment (pan3793) [inline docs/deployment/settings.md:208]: The generated doc will show the default value.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiServer.scala [hunk 4]
```diff
@@ -84,8 +86,15 @@ class KyuubiServer(name: String) extends Serverable(name) {
 
   override val backendService: AbstractBackendService = new KyuubiBackendService()
 
-  override val frontendServices: Seq[AbstractFrontendService] = Seq(
-    new KyuubiThriftBinaryFrontendService(this))
+  override lazy val frontendServices: Seq[AbstractFrontendService] =
+    conf.get(FRONTEND_PROTOCOLS).map(FrontendProtocols.withName).map {
+      case THRIFT_BINARY => new KyuubiThriftBinaryFrontendService(this)
+      case REST =>
+        warn("REST frontend protocol is experimental, API may changes in the future.")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiServer.scala:93]: `changes` -> `change`?

**Review by yanghua — APPROVED** [review]
LGTM

**Comment by pan3793** [comment]
Thanks, merging to master

**Pull request closed without merging**